lib/puppetserver/ca/create_action.rb

```diff
@@ -0,0 +1,268 @@
+          @logger.err "    Can't create a cert for a certname when one already exists."
+          return false
+        else
+          @logger.err 'Error:'
+          @logger.err "    When certificate request submitted for #{certname}:"
+          @logger.err "      code: #{result.code}"
+          @logger.err "      body: #{result.body.to_s}" if result.body
+          return false
+        end
+      end
+
+      # Make an HTTP request to CA to sign the named certificates
+      # @param certname [Array] the name of the certificate to have signed
+      # @param settings [Hash] a hash of config settings
+      # @return [Boolean] the success of certificates being signed
+      def sign_cert(certname, settings)
+        client = http_client(settings)
+
+        url = client.make_ca_url(settings[:ca_server],
+                       settings[:ca_port],
+                       'certificate_status')
+
+        client.with_connection(url) do |connection|
+          url.resource_name = certname
+          body = JSON.dump({desired_state: 'signed'})
+          result = connection.put(body, url)
+          check_sign_result(result, certname)
+        end
+      end
+
+      def check_sign_result(result, certname)
+        case result.code
+        when '204'
+          @logger.inform "Successfully signed certificate request for #{certname}"
+          return true
+        else
+          @logger.err 'Error:'
+          @logger.err "    When signing request submitted for #{certname}:"
+          @logger.err "      code: #{result.code}"
+          @logger.err "      body: #{result.body.to_s}" if result.body
+          return false
+        end
+      end
+
+      # Make an HTTP request to fetch the named certificates from CA
+      # @param certname [Array] the name of the certificate to fetch
+      # @param settings [Hash] a hash of config settings
+      # @return [Boolean] the success of certificate being downloaded
+      def download_cert(certname, settings)
+        client = http_client(settings)
+        url = client.make_ca_url(settings[:ca_server],
+                       settings[:ca_port],
+                       'certificate')
+        client.with_connection(url) do |connection|
+          url.resource_name = certname
+          result = connection.get(url)
+          if downloaded = check_download_result(result, certname)
+            save_file(result.body, certname, settings[:certdir], "Certificate")
+            @logger.inform "Successfully downloaded and saved certificate #{certname} to #{settings[:certdir]}/#{certname}.pem"
+          end
+          downloaded
+        end
+      end
+
+      def check_download_result(result, certname)
+        case result.code
+        when '200'
+          return true
+        when '404'
+          @logger.err 'Error:'
+          @logger.err "    Certificate could not be found/created/saved for #{certname}"
```

Comment:
Maybe something more specific? I'm think something like, "Signed certificate '#{certname}' could not be found on the CA"?

Reply:
ok